tests/test_ncbi.py

```diff
@@ -0,0 +1,54 @@
+import pytest
+import types
+from fastapi.exceptions import HTTPException
+
+from semantic_search.ncbi import (
+    _medline_to_docs,
+    _safe_request,
+    _parse_medline,
+    _get_eutil_records,
+    uids_to_docs,
+    Settings,
+)
+
+settings = Settings()
+
+
+def test_invalid_uid_test():
+    with pytest.raises(HTTPException):
+        uid = ["93846392868"]
+        records = [{"id:": [uid]}]
+        _medline_to_docs(records)
+
+
+def test_safe_request():
+    eutils_params = {
+        "db": "pubmed",
+        "id": "9887103",
+        "retstart": 0,
+        "retmode": "xml",
+        "api_key": settings.ncbi_eutils_api_key,
+    }
+    url = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"
+    assert _safe_request(url, "POST", files=eutils_params).status_code == 200
+
+
+def test_get_eutil_records():
+    eutil = "efetch"
+    id = "9887103"
+    expected_response = '<generator object parse at 0x7f562cfe9900>'
```

Comment:
This is not correct. You need to figure out what the output actually is (after casting it to a list)

Reply:
The expected list is something like this: (I got it from printing the response of the function)
```python
[{'PMID': '9887103', 'OWN': 'NLM', 'STAT': 'MEDLINE', 'DCOM': '19990225', 'LR': '20190516', 'IS': '0890-9369 (Print) 0890-9369 (Linking)', 'VI': '13', 'IP': '1', 'DP': '1999 Jan 1', 'TI': 'The Drosophila activin receptor baboon signals through dSmad2 and controls cell proliferation but not patterning during larval development.', 'PG': '98-111', 'AB': 'The TGF-beta superfamily of growth and differentiation factors, including TGF-beta, Activins and bone morphogenetic proteins (BMPs) play critical roles in regulating the development of many organisms. These factors signal through a heteromeric complex of type I and II serine/threonine kinase receptors that phosphorylate members of the Smad family of transcription factors, thereby promoting their nuclear localization. Although components of TGF-beta/Activin signaling pathways are well defined in vertebrates, no such pathway has been clearly defined in invertebrates. In this study we describe the role of Baboon (Babo), a type I Activin receptor previously called Atr-I, in Drosophila development and characterize aspects of the Babo intracellular signal-transduction pathway. Genetic analysis of babo loss-of-function mutants and ectopic activation studies indicate that Babo signaling plays a role in regulating cell proliferation. In mammalian cells, activated Babo specifically stimulates Smad2-dependent pathways to induce TGF-beta/Activin-responsive promoters but not BMP-responsive elements. Furthermore, we identify a new Drosophila Smad, termed dSmad2, that is most closely related to vertebrate Smads 2 and 3. Activated Babo associates with dSmad2 but not Mad, phosphorylates the carboxy-terminal SSXS motif and induces heteromeric complex formation with Medea, the Drosophila Smad4 homolog. Our results define a novel Drosophila Activin/TGF-beta pathway that is analogous to its vertebrate counterpart and show that this pathway functions to promote cellular growth with minimal effects on patterning.', 'FAU': ['Brummel, T', 'Abdollah, S', 'Haerry, T E', 'Shimell, M J', 'Merriam, J', 'Raftery, L', 'Wrana, J L', "O'Connor, M B"], 'AU': ['Brummel T', 'Abdollah S', 'Haerry TE', 'Shimell MJ', 'Merriam J', 'Raftery L', 'Wrana JL', "O'Connor MB"], 'AD': ['Department of Molecular Biology and Biochemistry, University of California, Irvine, California 92697, USA.'], 'LA': ['eng'], 'SI': ['GENBANK/AF101386'], 'GR': ['GM47462/GM/NIGMS NIH HHS/United States'], 'PT': ['Journal Article', "Research Support, Non-U.S. Gov't", "Research Support, U.S. Gov't, P.H.S."], 'PL': 'United States', 'TA': 'Genes Dev', 'JT': 'Genes & development', 'JID': '8711660', 'RN': ['0 (Bone Morphogenetic Proteins)', '0 (DNA-Binding Proteins)', '0 (Drosophila Proteins)', '0 (RNA, Messenger)', '0 (Receptors, Growth Factor)', '0 (Smad2 Protein)', '0 (Trans-Activators)', 'EC 2.7.11.30 (Activin Receptors)', 'EC 2.7.11.30 (Activin Receptors, Type I)', 'EC 2.7.11.30 (Babo protein, Drosophila)'], 'SB': 'IM', 'MH': ['Activin Receptors', 'Activin Receptors, Type I', 'Amino Acid Sequence', 'Animals', 'Bone Morphogenetic Proteins/genetics', 'Cell Division', 'Cloning, Molecular', 'DNA-Binding Proteins/chemistry/*genetics', 'Drosophila/*embryology', 'Drosophila Proteins', 'Gene Expression Regulation, Developmental', 'In Situ Hybridization', 'Larva/genetics/*growth & development', 'Molecular Sequence Data', 'Phosphorylation', 'RNA, Messenger/genetics', 'Receptors, Growth Factor/*genetics/metabolism', 'Sequence Alignment', 'Sequence Analysis, DNA', 'Signal Transduction/*physiology', 'Smad2 Protein', 'Trans-Activators/chemistry/*genetics', 'Wings, Animal/growth & development'], 'PMC': 'PMC316373', 'EDAT': '1999/01/14 00:00', 'MHDA': '1999/01/14 00:01', 'CRDT': ['1999/01/14 00:00'], 'PHST': ['1999/01/14 00:00 [pubmed]', '1999/01/14 00:01 [medline]', '1999/01/14 00:00 [entrez]'], 'AID': ['10.1101/gad.13.1.98 [doi]'], 'PST': 'ppublish', 'SO': 'Genes Dev. 1999 Jan 1;13(1):98-111. doi: 10.1101/gad.13.1.98.'}]
```

Now I made that equal to the `expected`. Then I made `actual` equal to what you had said (`list(_get_eutil_records(eutil,id))`). However, this still triggers an error on the `pytest` test.  The error looks like this:
```python
>       assert actual_response == expected_response
E       assert [{'<?xm': ['v...icle>']}, ...] == [{'AB': 'The ...', ...], ...}]
E         At index 0 diff: {'<?xm': ['version="1.0" ?>'], '<!DO': ['YPE PubmedArticleSet PUBLIC "-//NLM//DTD PubMedArticle, 1st January 2019//EN" "https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_190101.dtd">'], '<Pub': ['dArticleSet>', 'dArticle>'], '': ['edlineCitation Status="MEDLINE" Owner="NLM">', '  <PMID Version="1">9</PMID>', '  <DateCompleted>', '      <Year>1976</Year>', '      <Month>01</Month>', '      <Day>23</Day>', '  </DateCompleted>', '  <DateRevised>', '      <Year>2019</Year>', '      <Month>06</Month>', '      <Day>23</Day>', '  </DateRevised>', '  <Article PubM...
E         
E         ...Full output truncated (3 lines hidden), use '-vv' to show

tests/test_ncbi.py:42: AssertionError
```

Reply:
Yeah, so this means `actual` does not equal expected. Either `expected` is wrong or actual is wrong. It is helpful to run pytest with the `-vv` argument to get more information on where they differ

Reply:
I found the actual value and replaced it and now it seems to work. I was getting the wrong list in the beginning. 👍

Reply:
Cool! Is it easy to update the other unit tests now?